core/jobs/batch_jobs/exp_version_history_computation_job.py

```diff
@@ -0,0 +1,675 @@
+                create the version history model.
+
+        Returns:
+            ExplorationVersionHistoryModel. The updated version history model.
+        """
+        # If the model does not exist, create it with the data from the
+        # reverted model. Otherwise, just update the data of the already
+        # existing model with the data from the reverted model.
+        if current_vh_model is None:
+            current_vh_model = exp_models.ExplorationVersionHistoryModel(
+                id=exp_models.ExplorationVersionHistoryModel.get_instance_id(
+                    exp_id, current_version
+                ),
+                exploration_id=exp_id,
+                exploration_version=current_version,
+                state_version_history=revert_to_vh_model.state_version_history,
+                metadata_last_edited_version_number=(
+                    revert_to_vh_model.metadata_last_edited_version_number
+                ),
+                metadata_last_edited_committer_id=(
+                    revert_to_vh_model.metadata_last_edited_committer_id
+                ),
+                committer_ids=revert_to_vh_model.committer_ids
+            )
+        else:
+            current_vh_model.state_version_history = (
+                revert_to_vh_model.state_version_history
+            )
+            current_vh_model.metadata_last_edited_version_number = (
+                revert_to_vh_model.metadata_last_edited_version_number
+            )
+            current_vh_model.metadata_last_edited_committer_id = (
+                revert_to_vh_model.metadata_last_edited_committer_id
+            )
+            current_vh_model.committer_ids = revert_to_vh_model.committer_ids
+
+        return current_vh_model
+
+    def _check_for_revert_commit(
+        self, change_list: List[exp_domain.ExplorationChange]
+    ) -> Optional[int]:
+        """Checks if revert commit is present in the change list and returns
+        the version number (if present).
+
+        Args:
+            change_list: list(ExplorationChange). The list of changes to check.
+
+        Returns:
+            Optional[int]. The revert version number (if present) or None.
+        """
+        for change in change_list:
+            if change.cmd == feconf.CMD_REVERT_COMMIT:
+                return change.version_number
+        return None
+
+    def _create_version_history_models(
+        self, model_group: FormattedModelGroupDict
+    ) -> Tuple[str, List[exp_models.ExplorationVersionHistoryModel]]:
+        """Creates the version history models for a particular exploration.
+
+        Args:
+            model_group: FormattedModelGroupDict. The formatted model group.
+
+        Returns:
+            Tuple[str, List[exp_models.ExplorationVersionHistoryModel]].
+            The exploration id along with the created version history models.
+        """
+        with datastore_services.get_ndb_context():
```

Comment:
Why do we need this?

Reply:
Without this, I face the error `NDB calls must be made in context established by google.cloud.ndb.Client.context.`

It is because new models are being created in this step.